### PR TITLE
Validate email message size is within SES limits

### PR DESCRIPTION
**Pull request by CrystalPea**
SES rejects email messages bigger than 10485760 bytes
(just over 10 MB per message (after base64 encoding)):
https://docs.aws.amazon.com/ses/latest/DeveloperGuide/quotas.html#limits-message

Base64 is apparently wasteful because we use just 64 different
values per byte, whereas a byte can represent 256 different characters.
That is, we use bytes (which are 8-bit words) as 6-bit words. There is
a waste of 2 bits for each 8 bits of transmission data. To send
three bytes of information (3 times 8 is 24 bits), you need to use
four bytes (4 times 6 is again 24 bits). Thus the base64 version
of a file is 4/3 larger than it might be.
So we use 33% more storage than we could.
from:
https://lemire.me/blog/2019/01/30/what-is-the-space-overhead-of-base64-encoding/

So the max size of our message should be 7.5 MB == 7500000 bytes
before base64 encoding. We could techically make the limit a bit bigger,
since the limit is 10.48MB, not 10MB, but I decided to leave a margin,
because I don't know if subject line, recipient addresses etc.
count toward the limit. 7.5MB worth of text would be an extremely
long email anyway. Not very good from user experience perspective.

I doubt anything but some weird test emails would reach that limit.

Pivotal ticket: https://www.pivotaltracker.com/story/show/175115179 
Connected API PR: https://github.com/alphagov/notifications-api/pull/2992

### CODE REVIEW

File: tests/test_template_types.py [hunk 1]
```diff
@@ -1994,7 +1994,7 @@ def test_broadcast_message_too_long(template_class, extra_characters, expected_t
     (LetterPreviewTemplate, 'letter', {}),
     (LetterImageTemplate, 'letter', {'image_url': 'foo', 'page_count': 1}),
 ])
-def test_non_sms_ignores_message_too_long(template_class, template_type, kwargs):
+def test_message_too_long_limit_bigger_or_nonexistent_for_non_sms_templates(template_class, template_type, kwargs):
```
[the remark this answers is not in the document]
Comment (CrystalPea) [reply tests/test_template_types.py:1997]: This test is testing exactly what you wrote, yeah ☺️ . I did not add it, I just changed the name so it is more apt now we started checking size of email messages.

I'd say those tests are a bit different:
`test_message_too_long_limit_bigger_or_nonexistent_for_non_sms_templates` tests that email and letters slightly bigger than sms character limit pass validation

`test_message_too_long_for_an_email_message_within_limits` tests that emails that are pretty big (several MB), but are under the the size limit for emails pass validation.

File: notifications_utils/template.py [hunk 2]
```diff
@@ -535,6 +536,26 @@ def html_body(self):
...
+            SES rejects email messages bigger than 10485760 bytes (just over 10 MB per message (after base64 encoding)):
+            https://docs.aws.amazon.com/ses/latest/DeveloperGuide/quotas.html#limits-message
+
+            Base64 is apparently wasteful because we use just 64 different values per byte, whereas a byte can represent
+            256 different characters. That is, we use bytes (which are 8-bit words) as 6-bit words. There is
+            a waste of 2 bits for each 8 bits of transmission data. To send three bytes of information
+            (3 times 8 is 24 bits), you need to use four bytes (4 times 6 is again 24 bits). Thus the base64 version
+            of a file is 4/3 larger than it might be. So we use 33% more storage than we could.
+            https://lemire.me/blog/2019/01/30/what-is-the-space-overhead-of-base64-encoding/
+
```
[the remark this answers is not in the document]
Comment (CrystalPea) [reply notifications_utils/template.py:554]: Thank you 😊

File: tests/test_template_types.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (idavidmcdonald) [reply tests/test_template_types.py:1997]: I think I'd argue that I feel that is still testing the same thing, that if you have an email with length under 10MB that it is marked as not too long (it doesn't matter if the length of the email you are testing is 1 byte long, 1000 bytes, 1 million bytes, as long as it is under the length for email we care.

but I don't feel too strongly and it definitely won't harm us to have an extra test in so no changes needed but thanks for explaining. 

What I might suggest though is changing the name of the test to something like `test_message_too_long_limit_is_bigger_than_the_sms_limit_for_non_sms_templates`. I think the key bit in the name is that the limit is bigger **than the sms limit**.

**Review by idavidmcdonald — APPROVED** [review]
Nice work!

**Comment by quis** [comment]
Does this account for:
- the size taken up by sending SES the content of the email twice in each request (once as plain text and once as HTML)?
- the size taken up by the markup in the HTML version of the email?

Would also be worth checking if this extra check will have any significant impact on the performance of processing a big CSV file with a wordy template.

**Comment by idavidmcdonald** [comment]
I would like to withdraw my approval for the moment as Chris has asked some very very good questions!

File: tests/test_template_types.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (CrystalPea) [reply tests/test_template_types.py:1997]: I like the idea of changing the name, will do 👍

**Comment by CrystalPea** [comment]
Hey @quis those are good points.

> the size taken up by sending SES the content of the email twice in each request (once as plain text and once as HTML)?

Ahh, good point. I checked an email with some mark-up and html and plain looked the same for it, and had same size. When do you think they would differ? Probably when logo and banner is present?

> Would also be worth checking if this extra check will have any significant impact on the performance of processing a big CSV file with a wordy template.
That's something worth checking 👍

**Comment by quis** [comment]
Yeah, just looking at the content of the email won’t give you the HTML, banner and other formatting. You need to look at the rendered template instead, by calling `str` on it. Here’s a little example:
```python
import sys
from notifications_utils.template import PlainTextEmailTemplate, HTMLEmailTemplate

template_dict = {'content': 'Hello', 'subject': 'Test', 'template_type': 'email'}
plain, html = PlainTextEmailTemplate(template_dict), HTMLEmailTemplate(template_dict)

sys.getsizeof(str(plain))
>>> 119

sys.getsizeof(str(html))
>>> 11044
```

**Comment by CrystalPea** [comment]
@quis ahh, that makes sense, thanks!

**Review by quis — APPROVED** [review]
Nice, thanks for persevering with this and sorry it took so long to get round to showing you the profiling stuff

**Pull request merged**